### PR TITLE
Improve deprecation details reported via Problems API

### PR DESCRIPTION
**Pull request by hegyibalint**

### CODE REVIEW

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java [hunk 1]
```diff
@@ -346,5 +353,32 @@ public void run() {
...
+    }
+
+    private static void transformIntoProblem(final DeprecatedFeatureUsage usage) {
+        InternalProblems problems = ProblemsProgressEventEmitterHolder.get();
+        Action<InternalProblemSpec> problemBuilderSpec = new Action<InternalProblemSpec>() {
+            @Override
+            public void execute(InternalProblemSpec builder) {
+                InternalProblemSpec internalProblemSpec = builder
+                    .label(usage.getSummary())
+                    // FIXME: Not sure
```
Comment (donat) [inline platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java:367]:
```suggestion
```

**Comment by reinsch82** [comment]
@bot-gradle test this

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java [hunk 2]
```diff
@@ -346,5 +350,28 @@ public void run() {
 
     private synchronized static void nagUserWith(DeprecatedFeatureUsage usage) {
         DEPRECATED_FEATURE_HANDLER.featureUsed(usage);
+
+        reportAsProblem(usage);
+    }
+
+    private static void reportAsProblem(final DeprecatedFeatureUsage usage) {
+        InternalProblemReporter reporter = ProblemsProgressEventEmitterHolder.get().getInternalReporter();
+        reporter.report(reporter.create(new Action<InternalProblemSpec>() {
```
Comment (donat) [inline platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java:359]: ❌ `reporter.reporting(...)`

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java [hunk 3]
```diff
@@ -67,8 +74,8 @@ protected Integer initialValue() {
 
     private static final LoggingDeprecatedFeatureHandler DEPRECATED_FEATURE_HANDLER = new LoggingDeprecatedFeatureHandler();
 
-    public synchronized static void init(ProblemDiagnosticsFactory problemDiagnosticsFactory, WarningMode warningMode, BuildOperationProgressEventEmitter buildOperationProgressEventEmitter, Problems problemsService) {
-        DEPRECATED_FEATURE_HANDLER.init(problemDiagnosticsFactory, warningMode, buildOperationProgressEventEmitter, problemsService);
+    public synchronized static void init(ProblemDiagnosticsFactory problemDiagnosticsFactory, WarningMode warningMode, BuildOperationProgressEventEmitter buildOperationProgressEventEmitter) {
+        DEPRECATED_FEATURE_HANDLER.init(problemDiagnosticsFactory, warningMode, buildOperationProgressEventEmitter);
```
Comment (donat) [inline platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java:78]: ❓ Why did we change to the static problems service accessor here?

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java [hunk 4]
```diff
@@ -89,33 +77,9 @@ public void featureUsed(final DeprecatedFeatureUsage usage) {
                 error = new GradleException(WARNING_SUMMARY + " " + DefaultGradleVersion.current().getNextMajorVersion().getVersion());
             }
         }
-        if (problemsService != null) {
-            InternalProblemReporter reporter = ((InternalProblems) problemsService).getInternalReporter();
```
Comment (donat) [inline platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java:93]: I may miss something here. Was this the wrong place to do the actual problem report?

File: platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r86/DeprecationLoggingProblemEventsCrossVersionTest.groovy [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+        when:
+        def listener = new ProblemEventListener()
+        withConnection { connection ->
+            connection.newBuild()
+                .forTasks(":help")
+                .addProgressListener(listener)
+                .setStandardError(System.err)
+                .setStandardOutput(System.out)
+                .addArguments("--warning-mode", "all")
```
Comment (donat) [inline platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r86/DeprecationLoggingProblemEventsCrossVersionTest.groovy:41]: IIUC Problem reports will appear even wo/this configuration.
```suggestion
```

**Review by donat — CHANGES_REQUESTED** [review]
❌ Name of the PR indicates that it's WIP. Also, there's no description.

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java (hunk 3 above)
Comment (reinsch82) [reply platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java:78]: It's not necessary anymore. All the usage is now in this file by using `ProblemsProgressEventEmitterHolder.get()`

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java (hunk 4 above)
Comment (hegyibalint) [reply platforms/core-runtime/logging/src/main/java/org/gradle/internal/featurelifecycle/LoggingDeprecatedFeatureHandler.java:93]: As far as I remember (it wasn't yesterday when I wrote this), my motivation was to put the problem reporting to the highest point in the hierarchy possible. This location might be right functionality-wise, but I felt it was a more logical, better discoverable place if placed next to the `nag...` calls as possible.

File: platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java (hunk 3 above)
Comment (hegyibalint) [reply platforms/core-runtime/logging/src/main/java/org/gradle/internal/deprecation/DeprecationLogger.java:78]: Painful question, but just to make sure: can we deprecate stuff from a worker?

**Comment by hegyibalint** [comment]
❓ I know how painful this question is, but: can we deprecate stuff from workers?

File: platforms/core-runtime/logging/src/test/groovy/org/gradle/internal/deprecation/DeprecationMessagesTest.groovy [hunk 6]
```diff
@@ -44,9 +42,8 @@ class DeprecationMessagesTest extends Specification {
     def setup() {
         def diagnosticsFactory = new NoOpProblemDiagnosticsFactory()
 
-        def buildOperationProgressEventEmitter = Stub(BuildOperationProgressEventEmitter)
-        def problemEmitter = Stub(ProblemEmitter)
-        DeprecationLogger.init(diagnosticsFactory, WarningMode.All, buildOperationProgressEventEmitter, new DefaultProblems(problemEmitter))
+        def mock = Mock(BuildOperationProgressEventEmitter)
```
Comment (tresat) [inline platforms/core-runtime/logging/src/test/groovy/org/gradle/internal/deprecation/DeprecationMessagesTest.groovy:45]: 💅 I think it's better to create it with `Stub` if you're not interested in verifying interactions with a test double.

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/internal/DefaultExceptionContainer.java [hunk 7]
```diff
@@ -16,19 +16,20 @@
...
+    private final Failure exception;
 
-    public DefaultExceptionContainer(@Nullable RuntimeException exception) {
+    public DefaultExceptionContainer(@Nullable Failure exception) {
         this.exception = exception;
     }
 
     @Override
-    public RuntimeException getException() {
+    public Failure getException() {
```
Comment (tresat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/internal/DefaultExceptionContainer.java:32]: 🤔 Can the names of these methods be changed to `getFailure` as well, as it is no longer returning an exception?

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/internal/DefaultExceptionContainer.java [hunk 8]
```diff
@@ -16,19 +16,20 @@
 
 package org.gradle.tooling.events.problems.internal;
 
+import org.gradle.tooling.Failure;
 import org.gradle.tooling.events.problems.ExceptionContainer;
 
 import javax.annotation.Nullable;
 
 public class DefaultExceptionContainer implements ExceptionContainer {
```
Comment (tresat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/internal/DefaultExceptionContainer.java:24]: 🤔 The name of these container types should probably also change to `FailureContainer`.

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/parameters/BuildProgressListenerAdapter.java [hunk 9]
```diff
@@ -793,8 +793,11 @@ private static List<ProblemDescriptor> toProblemDescriptors(Collection<InternalP
         return result.build();
     }
 
-    private static ExceptionContainer toExceptionContainer(@Nullable RuntimeException exception) {
-        return new DefaultExceptionContainer(exception);
+    private static @Nullable ExceptionContainer toExceptionContainer(@Nullable InternalFailure exception) {
```
Comment (tresat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/parameters/BuildProgressListenerAdapter.java:796]: 🤔 And the name of this method, too.

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/parameters/BuildProgressListenerAdapter.java [hunk 10]
```diff
@@ -1056,7 +1059,7 @@ private static Failure toFailure(InternalFailure origFailure) {
             toFailures(origFailure.getCauses()));
     }
 
-    private static List<AnnotationProcessorResult> toAnnotationProcessorResults(List<InternalAnnotationProcessorResult> protocolResults) {
+    private static List<AnnotationProcessorResult> toAnnotationProcessorResults(@Nullable List<InternalAnnotationProcessorResult> protocolResults) {
```
Comment (tresat) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/parameters/BuildProgressListenerAdapter.java:1062]: 🤔 Seems like a code smell to have a method that takes a collection parameter have that parameter marked `@Nullable`.  Shouldn't the client just pass an empty collection?

File: platforms/ide/problems-api/src/main/java/org/gradle/internal/problems/DefaultProblemDiagnosticsFactory.java [hunk 11]
```diff
@@ -120,7 +130,7 @@ public DefaultProblemStream() {
             remainingStackTraces.set(maxStackTraces);
         }
 
-        public ProblemDiagnostics forCurrentCaller(@Nullable Throwable exception) {
+        public ProblemDiagnostics forCurrentCaller(@Nullable final Throwable exception) {
```
Comment (tresat) [inline platforms/ide/problems-api/src/main/java/org/gradle/internal/problems/DefaultProblemDiagnosticsFactory.java:133]: 💅 The `final` on this parameter here stands out, should it?

File: platforms/core-runtime/launcher/src/main/java/org/gradle/tooling/internal/provider/LauncherServices.java [hunk 12]
```diff
@@ -214,6 +215,12 @@ BuildSessionActionExecutor createActionExecutor(
     }
 
     static class ToolingBuildTreeScopeServices {
+
+        ProblemStream createProblemStream(StartParameter startParameter, ProblemDiagnosticsFactory problemDiagnosticsFactory) {
+            return startParameter.getWarningMode().shouldDisplayMessages()
+                ? problemDiagnosticsFactory.newUnlimitedStream()
+                : problemDiagnosticsFactory.newStream();
```
Comment (tresat) [inline platforms/core-runtime/launcher/src/main/java/org/gradle/tooling/internal/provider/LauncherServices.java:222]: 💅 This reads weird to me.  Seems like the `shouldDisplayMessages()` value is _not_ controlling whether or not to display messages, but rather _how many_ messages to display.

**Review by tresat — APPROVED** [review]
I think it's worth adjusting some of the naming to align better with the new types used, but looks good overall to me.

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/parameters/BuildProgressListenerAdapter.java (hunk 10 above)
Comment (reinsch82) [reply platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/parameters/BuildProgressListenerAdapter.java]: I agree, but the source this is filled from is @Nullable, so to get rid of the warning I can either add it here or make sure not returning `null` there is not breaking anything :(

File: platforms/core-runtime/launcher/src/main/java/org/gradle/tooling/internal/provider/LauncherServices.java (hunk 12 above)
Comment (reinsch82) [reply platforms/core-runtime/launcher/src/main/java/org/gradle/tooling/internal/provider/LauncherServices.java:222]: Good point, I didn't want to mess with this though it is just moving the logic for this here so I can reuse the `ProblemStream` globally.

**Review by donat — APPROVED** [review]
Fyi I did a bit of cleanup: https://github.com/gradle/gradle/pull/27051/commits/5a88da5431c3c1c0ec071c3517126436c2682340. Let me know if it makes sense.

Otherwise all good!

**Comment by donat** [comment]
@bot-gradle merge

**Comment by reinsch82** [comment]
@bot-gradle test ACT APT

**Comment by reinsch82** [comment]
@bot-gradle test ACT APT

**Comment by reinsch82** [comment]
@bot-gradle test PT

**Comment by reinsch82** [comment]
@bot-gradle test PT

**Comment by reinsch82** [comment]
@bot-gradle test APT ACT

**Comment by reinsch82** [comment]
@bot-gradle merge

**Pull request merged**